File: android/src/main/java/com/stario10module/StarSpoolJobStatusListWrapper.kt
package com.stario10module

import com.facebook.react.bridge.*
import com.starmicronics.stario10.StarIO10Exception

class StarSpoolJobStatusListWrapper internal constructor(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
    override fun getName(): String {
        return "StarSpoolJobStatusListWrapper"
    }

    @ReactMethod
    fun getJobStatusList(identifier: String, promise: Promise) {
        val statusList = InstanceManager.get(identifier)

        if (statusList is List<*>) {
            promise.resolve(StarIO10ValueConverter.toWritableArray(statusList))
        }
        else {
            promise.reject(StarIO10Exception("Identifier error"))
        }
    }

    @ReactMethod
    fun dispose(identifier: String, promise: Promise) {
        InstanceManager.remove(identifier)
        promise.resolve(0)
    }
}

File: ios/StarPrinter/StarSpoolJobStatusListWrapper.m
#import "StarSpoolJobStatusListWrapper.h"
#import "StarObjectManager.h"
#import "StarIO10ValueConverter.h"
#import <React/RCTLog.h>
@import StarIO10;


@interface StarSpoolJobStatusListWrapper()

@property(assign, nonatomic) StarObjectManager *objManager;

@end

@implementation StarSpoolJobStatusListWrapper

- (instancetype)init
{
    self = [super init];
    if (self) {
        _objManager = StarObjectManager.sharedManager;
    }
    return self;
}

+ (BOOL)requiresMainQueueSetup
{
    return NO;
}

RCT_EXPORT_MODULE()

RCT_REMAP_METHOD(dispose,
                 disposeWithObjectIdentifier:(nonnull NSString *)objID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    [_objManager remove:objID];
 
    resolve(nil);
}

RCT_REMAP_METHOD(getJobStatusList,
                 getJobStatusListWithObjectIdentifier:(nonnull NSString *)objID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    NSArray<STARIO10StarSpoolJobStatus *> *statusList = [_objManager getObject:objID];
    
    if (statusList == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    resolve([StarIO10ValueConverter toStarSpoolJobStatusDictionaryArray:statusList]);
}

@end
